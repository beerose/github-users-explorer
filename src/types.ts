import * as github from "./api/githubClient";

export type User = Pick<
  github.User,
  | "avatar_url"
  | "bio"
  | "blog"
  | "company"
  | "html_url"
  | "location"
  | "login"
  | "name"
>;

export type Repository = Pick<
  github.Repository,
  "html_url" | "description" | "stargazers_count" | "name"
>;

export type ErrorMessage = { message: string };

export const isErrorMessage = (x: unknown): x is ErrorMessage =>
  typeof x === "object" &&
  x !== null &&
  "message" in x &&
  typeof (x as Record<string, unknown>).message === "string";
