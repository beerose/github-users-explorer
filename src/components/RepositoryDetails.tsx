import React from "react";

import { styled } from "../theme";

import { StarIcon } from "./StarIcon";
import { Repository } from "../types";

const RepositoryLink = styled.a`
  display: flex;
  flex-direction: column;
  border: 1px solid ${({ theme }) => theme.colors.text09};
  border-radius: 3px;
  width: 15em;
  min-height: 10em;
  background: white;

  text-decoration: none;
  color: inherit;

  margin: 1em;
  padding: 0.5em 1em;

  :hover {
    border: 1px solid ${({ theme }) => theme.colors.text09};
    box-shadow: 0 0 10px 1px ${({ theme }) => theme.colors.text09};
  }

  h3 {
    margin: 0.5em 0;
  }

  p {
    margin: 10px 0;
    overflow: hidden;
    text-overflow: ellipsis;
    color: ${({ theme }) => theme.colors.text05};
  }
`;

const StarsDetails = styled.div`
  align-self: flex-end;

  display: flex;
  align-items: center;

  svg {
    margin-right: 0.2em;
    width: 16px;
  }
`;

type RepositoryDetailsProps = {
  repository: Repository;
};
export const RepositoryDetails: React.FC<RepositoryDetailsProps> = ({
  repository,
}) => {
  return (
    <RepositoryLink href={repository.html_url}>
      <StarsDetails>
        <StarIcon />
        {repository.stargazers_count}
      </StarsDetails>
      <h3>{repository.name}</h3>
      <p>{repository.description}</p>
    </RepositoryLink>
  );
};
