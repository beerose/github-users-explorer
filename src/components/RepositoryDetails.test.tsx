import React from "react";
import "@testing-library/jest-dom/extend-expect";

import { render } from "../../test/testUtils";
import { Repository } from "../types";

import { RepositoryDetails } from "./RepositoryDetails";

const repo: Repository = {
  name: "typescript-build2016-demos",
  html_url: "https://github.com/ahejlsberg/typescript-build2016-demos",
  description: "Collection of TypeScript demo projects form //build 2016",
  stargazers_count: 64,
};

describe("RepositoryDetails", () => {
  it("links to the GitHub account", () => {
    const { getByRole } = render(<RepositoryDetails repository={repo} />);

    expect(getByRole("link")).toHaveAttribute("href", repo.html_url);
  });

  it("matches snapshots", () => {
    const { asFragment } = render(<RepositoryDetails repository={repo} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
