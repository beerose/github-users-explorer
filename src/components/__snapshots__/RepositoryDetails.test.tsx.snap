// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryDetails matches snapshots 1`] = `
<DocumentFragment>
  <a
    class="css-19p57ut"
    href="https://github.com/ahejlsberg/typescript-build2016-demos"
  >
    <div
      class="css-1p91yv2"
    >
      <svg
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M12 2L15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2z"
        />
      </svg>
      64
    </div>
    <h3>
      typescript-build2016-demos
    </h3>
    <p>
      Collection of TypeScript demo projects form //build 2016
    </p>
  </a>
</DocumentFragment>
`;
